multi3.c

```diff
@@ -0,0 +1,206 @@
+
+#include "utils.h"
+
+// Global because they're heavily shared and don't change throughout run, so its
+// a good idea
+
+
+int p, d, maxfe, n, b, k;
+double (*f)(double x[]);
+
+typedef struct {
+  int start;
+  int end;
+  int iters;
+} params;
+
+double *solutions;
+double *all_fits;
+
+// Shared vars, lock plz
+double *global_best_sol, *global_worst_sol;
+double global_min_fit = INFINITY;
+double global_max_fit = 0;
+
+void *jaya_sub(void *point) {
+  params *par = (params *)point;
+  printf("start:%i end:%i iter:%i\n", par->start, par->end, par->iters);
+
+  double *r1 = malloc(sizeof(double) * d);
+  double *r2 = malloc(sizeof(double) * d);
+  double *xi = malloc(sizeof(double) * d);
+  double *best_sol = malloc(sizeof(double) * d);
+  double *worst_sol = malloc(sizeof(double) * d);
+
+  double min_fit = INFINITY, max_fit = 0;
+  double fit, x, new_fit, old_fit;
+  int min_idx = -1, max_idx = -1;
+  int iter, i, j;
+
+  // determining best and worst sol at the start
+  for (i = par->start; i < par->end; i++) {
+    fit = f(solutions + (i * d));
+    all_fits[i] = fit;
+
+    if (fit < min_fit) {
+      min_fit = fit;
+      min_idx = i;
+      best_sol = solutions + (i * d);
+    }
+
+    if (fit > max_fit) {
+      max_fit = fit;
+      max_idx = i;
+      worst_sol = solutions + (i * d);
+    }
+  }
+
+  for (iter = 0; iter < par->iters; iter++) {
+    max_fit = 0;
+    // update the solutions
+    // xi' = xi + r1 * (best-xi) - r2 * (worst -xi)
+    for (i = par->start; i < par->end; i++) {
+      rand_matrix(r1, 1, d, 0, 1);
+      rand_matrix(r2, 1, d, 0, 1);
+
+
+      // Create the new solution xi
+      for (j = 0; j < d; j++) {
+        x = solutions[i * d + j];
+        xi[j] = x + r1[j] * (best_sol[j] - x) - r2[j] * (worst_sol[j] - x);
+      }
+
+      old_fit = all_fits[i];
+      new_fit = f(xi);
+
+      // If new one is better, use it instead
+      if (new_fit < old_fit) {
+        if (new_fit < min_fit) {
+          min_fit = new_fit;
+          min_idx = i;
+        }
+
+        all_fits[i] = new_fit;
+
+        for (j = 0; j < d; j++)
+          solutions[i * d + j] = xi[j];
+      }
+
+      if (all_fits[i] > max_fit) {
+        max_fit = all_fits[i];
+        max_idx = i;
+      }
+
+    }
+
+    best_sol = solutions + (min_idx * d);
+    worst_sol = solutions + (max_idx * d);
+
+    // TODO: if you are doing this add a mutex
+    // if (min_fit < global_min_fit) {
+    //   global_min_fit = min_fit;
+    // }
+#if DEBUG
+    // assert(f(best_sol) == min_fit);
+    // assert(f(worst_sol) == max_fit);
+#endif
+  }
+  global_best_sol = best_sol;
+  pthread_exit(NULL);
+}
+
+double *jaya() {
+  int k = 6;
+  int sub_pop_unit = p / k;
+  int merge_times = k - 1;
+  int merged = 0;
+  int iter_per_merge = n / merge_times;
+
+  while (merged < k) {
+    pthread_t *threads = malloc(sizeof(pthread_t) * k);
+    params *pars = malloc(sizeof(params) * k);
+
+    int iter_per_thread = iter_per_merge / (k - merged);
```

Comment:
If you could write a comment explaining this more, its not clear on first glance, also more comments in general so its easier for TA to understand.